seqan strings do not work with gtest

I am trying to integrate a CMake file for SeqAn and google's gtest in order to prepare for a larger project.

I am getting the following error (abbreviated) when I try to make the tests.

error: ambiguous overload for ‘operator<<’ (operand types are ‘std::ostream {aka std::basic_ostream<char>}’ and ‘const seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> >’)
   *os << value;

I am trying this out on my GitHub page (https://github.com/pnrobinson/GTT), in case anybody wants to see the gory details.

I realise that seqan has its own unit testing framework, but I would love to use gtest -- is there any workaround?
thanks Peter

---

reported by @pnrobinson

I digged into it a bit and it apparently is the fault of those two ambiguous overload of `<<`

 https://github.com/seqan/seqan/blob/a0691b9d05eb1ee56f513a5642f745231d77798b/include/seqan/sequence/string_base.h#L2027-L2035

and

 https://github.com/google/googletest/blob/0fe96607d85cf3a25ac40da369db62bbee2939a5/googletest/include/gtest/gtest-printers.h#L194-L211

When I added this more specific function overload, s.t. the generic overload does not conflict anymore.

```c++
template <typename TValue, typename TSpec>
inline std::ostream &
operator<<(std::ostream & target,
           String<TValue, TSpec> const & source)
{
    auto it = directionIterator(target, Output());
    write(it, source);
    return target;
}
```
But, I got the following error:

```console
In file included from /seqan-src/include/seqan/sequence.h:105:0,
                 from /GTT/sequence.h:6,
                 from /GTT/main.cpp:2:
/seqan-src/include/seqan/sequence/string_base.h: In function ‘std::ostream& seqan::operator<<(std::ostream&, const seqan::String<TValue, TSpec>&)’:
/seqan-src/include/seqan/sequence/string_base.h:2042:15: error: there are no arguments to ‘directionIterator’ that depend on a template parameter, so a declaration of ‘directionIterator’ must be available [-fpermissive]
     auto it = directionIterator(target, Output());
               ^~~~~~~~~~~~~~~~~
```

I'm currently figuring out how to make `directionIterator` known.

--- a/include/seqan/sequence/string_base.h
+++ b/include/seqan/sequence/string_base.h
@@ -2034,6 +2034,30 @@ operator<<(TStream & target,
     return target;
 }
 
+// A specialization needed to avoid an ambiguous call combined with googletest.
+//
+// We use `std::basic_ostream<char>` instead of `std::ostream`, because we would
+// need to redefine a lot of CONCEPTS within seqan. Unfortunately,
+// `std::basic_ostream<char>` does not work, because `directionIterator` might not
+// be included yet (defined in <iter_stream.h>). Thus using `TChar` with the
+// restriction that it can only be `char`.
+//
+// We do not use the more generic `std::basic_ostream<TChar, Traits>` type,
+// because this would clash with other overloads of `<<` where `TSpec` within
+// String<TValue, ...> is more specific. E.g. in the case of `TSpec =
+// Journaled<...>`.
+//
+// https://github.com/seqan/seqan/issues/2182
+template <typename TChar, typename TValue>
+inline SEQAN_FUNC_ENABLE_IF(IsSameType<TChar, char>, std::basic_ostream<TChar> &)
+operator<<(std::basic_ostream<TChar> & target,
+           String<TValue> const & source)
+{
+    auto it = directionIterator(target, Output());
+    write(it, source);
+    return target;
+}
+
 // ----------------------------------------------------------------------------
 // Function operator>>() for streams.
 // ----------------------------------------------------------------------------

